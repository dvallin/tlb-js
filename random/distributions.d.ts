import { RandomSeed } from 'random-seed';
export interface Distribution {
    sample(): number;
}
export declare class Uniform implements Distribution {
    readonly rng: RandomSeed;
    constructor(seed: string);
    sample(): number;
}
export declare class Exponential implements Distribution {
    readonly distribution: Distribution;
    readonly lambda: number;
    static fromSeed(seed: string, lambda?: number): Exponential;
    private scale;
    constructor(distribution: Distribution, lambda?: number);
    sample(): number;
}
