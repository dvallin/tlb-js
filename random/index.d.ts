import { Distribution } from './distributions';
import { Rectangle } from '../geometry/rectangle';
import { Vector } from '../spatial';
export declare class Random {
    readonly distribution: Distribution;
    constructor(distribution: Distribution);
    floatBetween(minInclusive: number, maxExclusive: number): number;
    integerBetween(minInclusive: number, maxInclusive: number): number;
    decision(probability: number): boolean;
    weightedDecision(weights: number[]): number;
    pickN<T>(array: T[], n?: number): T[];
    pick<T>(array: T[]): T;
    pickIndex<T>(array: T[]): number;
    shuffle<T>(array: T[]): void;
    insideRectangle(rectangle: Rectangle): Vector;
    betweenVectors(a: Vector, b: Vector): Vector;
}
