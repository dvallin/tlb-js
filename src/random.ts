import * as seedrandom from "seedrandom"

let rng: seedrandom.prng

export function init(seed?: string): void {
    rng = seedrandom(seed)
}

export function uniformInteger(min: number = 0, max: number = 1): number {
    return (Math.abs(rng.int32()) % (max - min)) + min
}

export function bernoulli(p: number = 0.5): boolean {
    return rng.double() <= p
}

export function binomialNormed(p: number[]): number {
    let test = rng.double()
    return p.findIndex((t) => {
        if (test <= t) {
            return true
        }
        test -= t
        return false
    })
}

export function dualDecision(
    p00: number, [p01, p10]: number[], p11: number
): [boolean, boolean] {
    const decision = binomialNormed([p00, p01, p10, p11])
    return [
        Math.floor(decision / 2) % 2 === 1,
        decision % 2 === 1
    ]
}

export function ternaryDecision(
    p000: number, [p001, p010, p100]: number[], [p011, p101, p110]: number[], p111: number
): [boolean, boolean, boolean] {
    const decision = binomialNormed([p000, p001, p010, p011, p100, p101, p110, p111])
    return [
        Math.floor(decision / 4) % 2 === 1,
        Math.floor(decision / 2) % 2 === 1,
        decision % 2 === 1
    ]
}
