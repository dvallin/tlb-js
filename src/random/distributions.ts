import { create, RandomSeed } from "random-seed"

export interface Distribution {
    sample(): number
}

export class Uniform implements Distribution {
    public readonly rng: RandomSeed

    public constructor(seed: string) {
        this.rng = create(seed)
    }

    public sample(): number {
        return this.rng.random()
    }
}

export class Exponential implements Distribution {

    public static fromSeed(seed: string, lambda: number = 1) {
        return new Exponential(new Uniform(seed), lambda)
    }

    private scale: number

    public constructor(
        public readonly distribution: Distribution,
        public readonly lambda: number = 1
    ) {
        this.scale = 1 - Math.exp(-this.lambda)
    }

    public sample(): number {
        const u = this.distribution.sample()
        return -Math.log(1 - this.scale * u) / this.lambda
    }
}
