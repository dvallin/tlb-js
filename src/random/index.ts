import { Distribution } from './distributions'
import { Rectangle } from '../geometry/rectangle'
import { Vector } from '../spatial'

export class Random {
  public constructor(public readonly distribution: Distribution) {}

  public floatBetween(minInclusive: number, maxExclusive: number): number {
    return this.distribution.sample() * (maxExclusive - minInclusive) + minInclusive
  }

  public integerBetween(minInclusive: number, maxInclusive: number): number {
    const c = this.distribution.sample()
    return Math.floor(c * (maxInclusive - minInclusive + 1)) + minInclusive
  }

  public decision(probability: number): boolean {
    return this.distribution.sample() < probability
  }

  public weightedDecision(weights: number[]): number {
    const sum = weights.reduce((v, c) => v + c, 0)
    const pick = this.integerBetween(0, sum - 1)
    let agg = 0
    for (let i = 0; i < weights.length; i++) {
      agg += weights[i]
      if (pick < agg) {
        return i
      }
    }
    throw new Error('invalid input to weighted decision')
  }

  public pick<T>(array: T[]): T {
    return array[this.integerBetween(0, array.length - 1)]
  }

  public pickIndex<T>(array: T[]): number {
    return this.integerBetween(0, array.length - 1)
  }

  public shuffle<T>(array: T[]): void {
    for (let i = array.length - 1; i >= 0; i--) {
      const j = this.integerBetween(0, i)
      const t = array[i]
      array[i] = array[j]
      array[j] = t
    }
  }

  public insideRectangle(rectangle: Rectangle): Vector {
    return new Vector([this.integerBetween(rectangle.left, rectangle.right), this.integerBetween(rectangle.top, rectangle.bottom)])
  }
}
