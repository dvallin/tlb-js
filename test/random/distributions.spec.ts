import { Uniform, Exponential, Distribution } from "../../src/random/distributions"

const uniform = new Uniform("seed")

describe(Uniform.name, () => {

    samplesIntoUnitInterval(uniform)
})

describe(Exponential.name, () => {

    samplesIntoUnitInterval(Exponential.fromSeed("seed"))

    describe("lambda=0.1", () => {
        samplesIntoUnitInterval(Exponential.fromSeed("seed", 0.1))
        samplesNicely(new Exponential(uniform, 0.1), uniform)
    })
    describe("lambda=0.5", () => samplesNicely(new Exponential(uniform, 0.5), uniform))
    describe("lambda=1", () => samplesNicely(new Exponential(uniform, 1), uniform))
    describe("lambda=2", () => {
        samplesIntoUnitInterval(Exponential.fromSeed("seed", 2))
        samplesNicely(new Exponential(uniform, 2), uniform)
    })
})

function samplesIntoUnitInterval(distribution: Distribution) {
    it("samples into unit interval", () => {
        for (let i = 0; i < 1000; i++) {
            const r = distribution.sample()
            expect(r).toBeGreaterThan(0.0)
            expect(r).toBeLessThan(1.0)
        }
    })
}

function samplesNicely(distribution: Distribution, underlyingDistribution: Distribution): void {

    it("samples nicely", () => {
        let samples: number[] = []
        for (let i = 0; i <= 10; i++) {
            underlyingDistribution.sample = jest.fn().mockReturnValue(i / 10)
            samples.push(distribution.sample())
        }
        expect(samples).toMatchSnapshot()
    })
}
