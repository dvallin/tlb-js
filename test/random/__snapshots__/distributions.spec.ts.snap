// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exponential lambda=0.1 samples nicely 1`] = `
Array [
  -0,
  0.09561827108889881,
  0.19215966136159257,
  0.2896421686935464,
  0.38808432249774694,
  0.48750520486374443,
  0.5879244727581393,
  0.6893623813510894,
  0.7918398085380881,
  0.8953782807312824,
  1.0000000000000004,
]
`;

exports[`Exponential lambda=0.5 samples nicely 1`] = `
Array [
  -0,
  0.08028389750081043,
  0.16392581427868808,
  0.251218969828263,
  0.34249674946118525,
  0.43814039275967726,
  0.5385886145288046,
  0.6443497721247285,
  0.7560174333214578,
  0.8742905530525391,
  1,
]
`;

exports[`Exponential lambda=1 samples nicely 1`] = `
Array [
  -0,
  0.06529833599883364,
  0.1351602748368097,
  0.21027195642236868,
  0.29148693313768487,
  0.3798854930417225,
  0.4768628363884144,
  0.5842647781563711,
  0.7046054708796523,
  0.841434921259571,
  1,
]
`;

exports[`Exponential lambda=2 samples nicely 1`] = `
Array [
  -0,
  0.045217600345924584,
  0.0949347515143791,
  0.15014691032101748,
  0.21222144117418723,
  0.28310958475848635,
  0.3657349317918609,
  0.4647706948496654,
  0.5883925047211546,
  0.7529856459779107,
  1,
]
`;
